core/base/v7/inc/ROOT/TDirectory.h

```diff
@@ -148,35 +148,31 @@ class TDirectory {
   /// \throws TDirectoryTypeMismatch if the object stored under this name is of
   ///   a type that is not a derived type of `T`.
   template <class T>
-  std::shared_ptr<T> Get(const std::string& name) {
-    ContentMap_t::iterator idx = fContent.find(name);
+  decltype(auto) Get(const std::string& name) {
+    auto idx = fContent.find(name);
     if (idx != fContent.end()) {
-      // FIXME: implement upcast!
-      if (auto dep
-        = dynamic_cast<Internal::TDirectoryEntryPtr<T>*>(idx->second.get())) {
-        return dep->GetPointer();
-      }
+      if (idx->second.GetTypeInfo() == typeid(ToContentType_t<T>))
```

Comment:
Not all archs make unique_ptr-s identical across shared libraries. Can we have a
```
int ROOT::compare_typeinfo(const type_info& a, const type_info& b) {
  if (&a == &b) return 0;
  return strcmp(a.name(), b.name());
}
```
I don't know where to put thius, though - we probably need an include/ROOT/MixedBagOThings.h, where include/ROOT/Support.h is likely a better name ;-)

I can do this change if you prefer.